src/lib/draw-and-pick.js

```diff
@@ -206,11 +195,122 @@ export function pickLayers(gl, {
 }
 /* eslint-enable max-depth, max-statements */
 
+export function pickLayersByBoundingBox(gl, {
+  layers,
+  pickingFBO,
+  boundingBox: {x, y, width, height},
+  viewport
+}) {
+
+  // Convert from canvas top-left to WebGL bottom-left coordinates
+  // And compensate for pixelRatio
+  const pixelRatio = typeof window !== 'undefined' ?
+    window.devicePixelRatio : 1;
+  const deviceBox = {
+    x: Math.round(x * pixelRatio),
+    y: Math.round(gl.canvas.height - (y + height) * pixelRatio),
+    width: Math.round(width * pixelRatio),
+    height: Math.round(height * pixelRatio)
+  };
+
+  // For all other events, run picking process normally.
+  const pickResults = pickFromBufferByBoundingBox(gl, {
+    layers,
+    pickingFBO,
+    deviceBox
+  });
+
+  const baseInfo = createInfo([x, y], viewport);
+  baseInfo.pixelRatio = pixelRatio;
+
+  // Convert each result to pickingInfo object
+  const infos = {};
+
+  pickResults.filter(Boolean)
+  .forEach(({color, layer, objectIndex}) => {
+    // Get unique objects
+    const objectId = `${layer.id}:${objectIndex}`;
+
+    if (!infos[objectId]) {
+      const info = Object.assign({}, baseInfo, {
+        color,
+        index: objectIndex,
+        picked: true
+      });
+
+      infos[objectId] = getLayerPickingInfo({
+        layer,
+        info,
+        mode: 'query'
+      });
+    }
+  });
+
+  return Object.values(infos);
+}
+
+/**
+ * Pick at a specified pixel with a tolerance radius
+ * Returns the closest object to the pixel in shape `{color, layer, objectIndex}`
+ */
 function pickFromBuffer(gl, {
   layers,
   pickingFBO,
   deviceX,
-  deviceY
+  deviceY,
+  deviceRadius
+}) {
+
+  // Create a box of size `radius * 2 + 1` centered at [deviceX, deviceY]
+  const x = Math.max(0, deviceX - deviceRadius);
+  const y = Math.max(0, deviceY - deviceRadius);
+  const deviceBox = {
+    x,
+    y,
+    width: Math.min(pickingFBO.width, deviceX + deviceRadius) - x + 1,
+    height: Math.min(pickingFBO.height, deviceY + deviceRadius) - y + 1
+  };
+
+  const pickResults = pickFromBufferByBoundingBox(gl, {layers, pickingFBO, deviceBox});
+
+  // Traverse all pixels in picking results and find the one closest to the supplied
+  // [deviceX, deviceY]
+  let minDistanceToCenter = deviceRadius;
+  let closestResultToCenter = null;
+  /* eslint-disable max-depth */
+  for (let col = 0; col < deviceBox.width; col++) {
+    for (let row = 0; row < deviceBox.height; row++) {
+      const pickResult = pickResults[row * deviceBox.width + col];
+      if (pickResult) {
+        const dx = deviceBox.x + col - deviceX;
+        const dy = deviceBox.y + row - deviceY;
+        const d = Math.sqrt(dx * dx + dy * dy);
+        if (d <= minDistanceToCenter) {
```

Comment:
This comparison should be implemented using distance squared, instead of distance to save that super expansive sqrt operations